apphub/instance_detection/yolov5/yolov5_tf.py

```diff
@@ -0,0 +1,579 @@
+    def __init__(self, mscoco_ds):
+        self.mscoco_ds = mscoco_ds
+
+    def __len__(self):
+        return len(self.mscoco_ds)
+
+    def __getitem__(self, idx):
+        indices = [idx] + [random.randint(0, len(self) - 1) for _ in range(3)]
+        samples = [self.mscoco_ds[i] for i in indices]
+        return {
+            "image1": samples[0]["image"],
+            "bbox1": samples[0]["bbox"],
+            "image2": samples[1]["image"],
+            "bbox2": samples[1]["bbox"],
+            "image3": samples[2]["image"],
+            "bbox3": samples[2]["bbox"],
+            "image4": samples[3]["image"],
+            "bbox4": samples[3]["bbox"]
+        }
+
+
+class CombineMosaic(NumpyOp):
+    def forward(self, data, state):
+        image1, image2, image3, image4, bbox1, bbox2, bbox3, bbox4 = data
+        images = [image1, image2, image3, image4]
+        bboxes = [bbox1, bbox2, bbox3, bbox4]
+        images_new, boxes_new = self._combine_images_boxes(images, bboxes)
+        return images_new, boxes_new
+
+    def _combine_images_boxes(self, images, bboxes):
+        s = 640
+        yc, xc = int(random.uniform(320, 960)), int(random.uniform(320, 960))
+        images_new = np.full((1280, 1280, 3), fill_value=114, dtype=np.uint8)
+        bboxes_new = []
+        for idx, (image, bbox) in enumerate(zip(images, bboxes)):
+            h, w = image.shape[0], image.shape[1]
+            # place img in img4
+            if idx == 0:  # top left
+                x1a, y1a, x2a, y2a = max(xc - w, 0), max(yc - h, 0), xc, yc  # xmin, ymin, xmax, ymax (large image)
+                x1b, y1b, x2b, y2b = w - (x2a - x1a), h - (y2a - y1a), w, h  # xmin, ymin, xmax, ymax (small image)
+            elif idx == 1:  # top right
+                x1a, y1a, x2a, y2a = xc, max(yc - h, 0), min(xc + w, s * 2), yc
+                x1b, y1b, x2b, y2b = 0, h - (y2a - y1a), min(w, x2a - x1a), h
+            elif idx == 2:  # bottom left
+                x1a, y1a, x2a, y2a = max(xc - w, 0), yc, xc, min(s * 2, yc + h)
+                x1b, y1b, x2b, y2b = w - (x2a - x1a), 0, w, min(y2a - y1a, h)
+            elif idx == 3:  # bottom right
+                x1a, y1a, x2a, y2a = xc, yc, min(xc + w, s * 2), min(s * 2, yc + h)
+                x1b, y1b, x2b, y2b = 0, 0, min(w, x2a - x1a), min(y2a - y1a, h)
+            images_new[y1a:y2a, x1a:x2a] = image[y1b:y2b, x1b:x2b]
+            padw, padh = x1a - x1b, y1a - y1b
+            for x1, y1, bw, bh, label in bbox:
+                x1_new = np.clip(x1 + padw, x1a, x2a)
+                y1_new = np.clip(y1 + padh, y1a, y2a)
+                x2_new = np.clip(x1 + padw + bw, x1a, x2a)
+                y2_new = np.clip(y1 + padh + bh, y1a, y2a)
+                bw_new = x2_new - x1_new
+                bh_new = y2_new - y1_new
+                if bw_new * bh_new > 1:
+                    bboxes_new.append((x1_new, y1_new, bw_new, bh_new, label))
+        return images_new, bboxes_new
+
+
+class HSVAugment(NumpyOp):
```

Comment:
Our HueSaturationValue op didn't work for this? If not, should this one replace our current offering?

Reply:
From what I see, it looks like this implementation determines has different shift for different pixel value, rather than using the same shift throughout the image.  I think both Ops have their pros and cons, therefore no need to have one replacing the other.    This Op is used by official code, so I stick to this one for consistency.